asylo/platform/primitives/fortanix_edp/edp_runner/src/lib.rs

```diff
@@ -0,0 +1,258 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+extern crate aesm_client;
+extern crate enclave_runner;
+extern crate sgxs_loaders;
+extern crate sgxs;
+extern crate failure;
+extern crate num_bigint;
+extern crate sha2;
+
+use aesm_client::AesmClient;
+use enclave_runner::EnclaveBuilder;
+use enclave_runner::usercalls::{SyncStream, UsercallExtension};
+use sgxs_loaders::isgx::Device as IsgxDevice;
+use failure::{Error, ResultExt};
+
+use std::ffi::CStr;
+use std::io::{ErrorKind, Error as IoError, Result as IoResult, Write};
+use std::os::raw::{c_char, c_uchar};
+use std::sync::mpsc::{Sender, Receiver, channel};
+use std::sync::Mutex;
+use std::{slice, mem, thread};
+
+mod crypto;
+
+use self::crypto::{Hasher, PrivateKey};
+
+pub const ERR_NULL_OUTPUT_PARAM: u64 = 100;
+pub const ERR_OTHER: u64 = 101;
+pub const ERR_INVALID_ENCLAVE: u64 = 102;
+
+macro_rules! convert_error {
+    ( $err:ident, $err_size:ident, $body:expr ) => {{
+        match $body {
+            Ok(()) => 0,
+            Err(e) => {
+                if !$err.is_null() {
+                    let mut err = slice::from_raw_parts_mut($err, $err_size);
+                    write!(err, "{:?}\0", e).expect("failed to copy error message");
+                }
+                ERR_OTHER
+            }
+        }
+    }}
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn fortanix_edp_load_enclave(
+    path: *const c_char,
+    enclave: *mut *mut EdpEnclave,
+    err: *mut c_uchar,
+    err_size: usize,
+) -> u64 {
+    if enclave.is_null() {
+        return ERR_NULL_OUTPUT_PARAM;
+    }
+    convert_error!(err, err_size, (|| {
+        let path = CStr::from_ptr(path).to_str()?;
+        match EdpEnclave::load_enclave(path) {
+            Ok(edp_enclave) => {
+                let edp_enclave = Box::new(edp_enclave);
+                *enclave = Box::into_raw(edp_enclave);
+                Ok(())
+            },
+            Err(e) => Err(e)
+        }
+    })())
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn fortanix_edp_free_enclave(enclave: *mut EdpEnclave) {
+    if !enclave.is_null() {
+        let _ = Box::from_raw(enclave);
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn fortanix_edp_enclave_call(
+    enclave: *mut EdpEnclave,
+    selector: u64,
+    input: *const c_uchar,
+    input_size: usize,
+    output: *mut *mut c_uchar,
+    output_size: *mut usize,
+    err: *mut c_uchar,
+    err_size: usize,
+) -> u64 {
+    let enclave = match enclave.as_ref() {
```

Comment:
`enclave.as_ref().ok_or(ERR_INVALID_ENCLAVE)?`

Reply:
the function does not return a `Result`

Reply:
You can move it inside the closure